tests/python_on_whales/components/test_image.py

```diff
@@ -24,18 +30,20 @@ def test_image_repr():
     docker.image.remove(["busybox:1", "busybox:1.32"])
 
 
-def test_image_remove():
-    docker.image.pull("busybox:1", quiet=True)
-    docker.image.pull("busybox:1.32", quiet=True)
-    docker.image.remove(["busybox:1", "busybox:1.32"])
+@pytest.mark.parametrize("ctr_client", [docker_client, podman_client], indirect=True)
+def test_image_remove(ctr_client: DockerClient):
+    ctr_client.image.pull("busybox:1", quiet=True)
+    ctr_client.image.pull("busybox:1.32", quiet=True)
+    ctr_client.image.remove(["busybox:1", "busybox:1.32"])
 
 
-def test_image_save_load(tmp_path):
+@pytest.mark.parametrize("ctr_client", [docker_client], indirect=True)
+def test_image_save_load(ctr_client: DockerClient, tmp_path: Path):
```

Comment:
I don't like this (as I said [here](https://github.com/gabrieldemarmiesse/python-on-whales/pull/520#issuecomment-1880082873)), to me it seems like a strange way to say "this test only supports docker". What's more relevant here is that this test *doesn't yet work with podman*, and it would be good if that was indicated more explicitly (e.g. by causing an XFAIL).

Reply:
I believe it's possible to do this. We can provide an indicator to the DockerClient (similar to what you did in the previous PR) about what kind of engine DockerClient in plugged on, then we can do 

```python
@pytest.mark.parametrize("ctr_client", [docker_client], indirect=True)
def test_image_save_load(ctr_client: DockerClient, tmp_path: Path):
    if ctr_client.client_type == "podman":
        pytest.xfail("load and save doesn't work yet with podman but it should")
```

Reply:
Ok, so then every single testcase that *should* support both docker and podman will need the `@pytest.mark.parametrize("ctr_client", [docker_client, podman_client], indirect=True)` line, and every single one of those that's not yet working with podman will need those extra two lines to do the xfail? It just feels to me like a lot of boilerplate/distraction from what the tests are actually doing, but I get that it is very explicit.

Aside from that, I'm still against using `@pytest.mark.parametrize("ctr_client", [docker_client], indirect=True)` ("this test is marked to be paramterised indirectly useing the `ctr_client` fixture, and will take values [`docker_client`]") as a way to write "this test is only expected to work with docker".

Reply:
I don't mind verbosity and duplication in the unit tests. If that spills in the code that we're shipping, then that's an issue I do agree.

I do agree that it's a strange way to get the corresponding client. We could make a fixture that only gives you the docker client, but then the "docker" marker won't be applied to the function, and we'd rely on people thinking about it to have it applied manually. Or we could loop over all tests to check if they use the fixture and then apply the marker, which is some more custom code. I don't know if it's worth it, I'll take some time to think about it.

Reply:
This is where I think it's far more natural for the default to be that tests are parameterised, and then to explicitly mark tests that don't support certain runtimes such that they gets skipped/xfailed/marked/removed as required. It's more relevant/interesting when tests *don't* support a runtime than when they *do*, so better to add verbosity in the negative case (e.g. "doesn't support podman") rather than the positive case (parameterise with these runtimes).